File: python/p065.py
# 
# Solution to Project Euler problem 65
# Copyright (c) Project Nayuki. All rights reserved.
# 
# https://www.nayuki.io/page/project-euler-solutions
# https://github.com/nayuki/Project-Euler-solutions
# 


def compute():
	numer = 1
	denom = 0
	for i in reversed(range(100)):
		numer, denom = e_contfrac_term(i) * numer + denom, numer
	ans = sum(int(c) for c in str(numer))
	return str(ans)


def e_contfrac_term(i):
	if i == 0:
		return 2
	elif i % 3 == 2:
		return i // 3 * 2 + 2
	else:
		return 1


if __name__ == "__main__":
	print(compute())

File: haskell/p065.hs
{- 
 - Solution to Project Euler problem 65
 - Copyright (c) Project Nayuki. All rights reserved.
 - 
 - https://www.nayuki.io/page/project-euler-solutions
 - https://github.com/nayuki/Project-Euler-solutions
 -}

import qualified EulerLib


limit = 100
main = putStrLn (show ans)
ans = let
		func i end
			| i >= end = (1, 0)
			| otherwise = let (n, d) = func (i+1) end
			in ((eContFracTerm i) * n + d, n)
		(numer, _) = func 0 limit
	in EulerLib.digitSum numer

eContFracTerm 0 = 2
eContFracTerm i
	| mod i 3 == 2 = (div i 3) * 2 + 2
	| otherwise    = 1
